data_warehouse/services/coverage_statistics_services.py

```diff
@@ -0,0 +1,194 @@
+                按学院统计
+            total => int
+                覆盖的总人数（不是人次，是人数）
+        }
+        '''
+        data = {'titles': {}, 'ages': {}, 'departments': {}, 'total': -1}
+        user_ids = records.values_list('user', flat=True).distinct()
+        total_users_covered = len(user_ids)
+        ages_groupby, titles_groupby, department_groupby = {}, {}, {}
+        for uid in user_ids:
+            user = User.objects.get(pk=uid)
+            age = user.age
+            title = user.technical_title
+            department = str(user.department)
+            if department in department_groupby:
+                department_groupby[department]['coverage_count'] += 1
+            else:
+                department_groupby[department] = {}
+                department_groupby[department]['coverage_count'] = 1
+                department_groupby[department]['total_count'] = len(
+                    User.objects.filter(department=user.department))
+            if title in titles_groupby:
+                titles_groupby[title]['coverage_count'] += 1
+            else:
+                titles_groupby[title] = {}
+                titles_groupby[title]['coverage_count'] = 1
+                titles_groupby[title]['total_count'] = len(User.objects.filter(
+                    technical_title=title))
+            low, high, mid = 0, len(CoverageStatisticsService.AGES) - 1, -1
+            while low <= high:
+                mid = (low + high) // 2
+                small = CoverageStatisticsService.AGES[mid][1][0]
+                big = CoverageStatisticsService.AGES[mid][1][1]
+                if small <= age <= big:
+                    break
+                elif age < small:
+                    high = mid - 1
+                else:
+                    low = mid + 1
+            age_range_label = CoverageStatisticsService.AGES[mid][0]
+            age_range = tuple(CoverageStatisticsService.AGES[mid][1])
+            if age_range_label in ages_groupby:
+                ages_groupby[age_range_label]['coverage_count'] += 1
+            else:
+                ages_groupby[age_range_label] = {}
+                ages_groupby[age_range_label]['coverage_count'] = 1
+                ages_groupby[age_range_label]['total_count'] = len(
+                    User.objects.filter(age__range=age_range))
+        data['titles'] = titles_groupby
+        data['ages'] = ages_groupby
+        data['departments'] = department_groupby
+        data['total'] = total_users_covered
+        return data
+
+    # pylint: disable=R0912
+    @staticmethod
+    def get_traning_records(user, program_id, department_id=None,
+                            start_time=None, end_time=None):
+        '''根据项目ID和起始时间查询所有的培训活动的所有的培训记录（包含校内和校外活动)
+
+        Parameters
+        ----
+        user: User
+            执行此操作的用户
+
+        program_ids: int
+            培训项目ID，为None或者-999表示查询所有的培训项目。
+
+        department_id: int
+            需要导出培训项目所属的部门，为None则导出当前用户有权管理的院系的所有培训项目。
```

Comment:
`department_id=None` is only for school admin, and it means retrieve all records for all departments, department admin has to offer `department_id`, otherwise `BadRequest` will be raised, please make the comment clear.